osu.Game/Skinning/RulesetSkinProvidingContainer.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using JetBrains.Annotations;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets;
+
+namespace osu.Game.Skinning
+{
+    /// <summary>
+    /// A type of <see cref="SkinProvidingContainer"/> that provides access to the beatmap skin and user skin,
+    /// each transformed with the ruleset's own skin transformer individually.
+    /// </summary>
+    public class RulesetSkinProvidingContainer : SkinProvidingContainer
+    {
+        protected readonly Ruleset Ruleset;
+        protected readonly IBeatmap Beatmap;
+
+        protected override Container<Drawable> Content { get; }
+
+        public RulesetSkinProvidingContainer(Ruleset ruleset, IBeatmap beatmap, [CanBeNull] ISkin beatmapSkin)
+        {
+            Ruleset = ruleset;
+            Beatmap = beatmap;
+
+            InternalChild = new BeatmapSkinProvidingContainer(getRulesetTransformedSkin(beatmapSkin))
+            {
+                Child = Content = new Container
+                {
+                    RelativeSizeAxes = Axes.Both,
+                }
+            };
+        }
+
+        [Resolved]
+        private SkinManager skinManager { get; set; }
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            UpdateSkins();
+        }
+
+        protected override void OnSourceChanged()
+        {
+            UpdateSkins();
+            base.OnSourceChanged();
+        }
+
+        protected virtual void UpdateSkins()
+        {
+            SkinSources.Clear();
+
+            SkinSources.Add(getRulesetTransformedSkin(skinManager.CurrentSkin.Value));
+
+            // TODO: we also want to return a DefaultLegacySkin here if the current *beatmap* is providing any skinned elements.
+            if (skinManager.CurrentSkin.Value is LegacySkin && skinManager.CurrentSkin.Value != skinManager.DefaultLegacySkin)
+                SkinSources.Add(getRulesetTransformedSkin(skinManager.DefaultLegacySkin));
+
+            if (skinManager.CurrentSkin.Value != skinManager.DefaultSkin)
+                SkinSources.Add(getRulesetTransformedSkin(skinManager.DefaultSkin));
```

Comment:
This logic is currently duplicated in `SkinManager`. Is the intention to have this replace that logic, or is it deemed that this will be different at some point in the future?

Put another way, is there any reasons we're not exposing this from `SkinManager` itself?

Reply:
Because you've stated that it's weird to expose something like this there? I initially had it like that for the exact reason of not duplicating it.

I also left it both here and in `SkinManager` so that components outside the `Player` can also fall back to their legacy defaults.

Reply:
I'll bring the `CurrentSkinLayers` (as `CurrentSkinSources`) back for DRY purposes.

Reply:
Not sure it's required. Especially after you apply the other remaining review comment which will make them no longer match..?

Reply:
The skin sources remain the same in that other review comment, the difference is whether they always get transformed with the ruleset's skin transformer or some of them don't.

Reply:
Well my suggestion was to *remove* the default source and have that fall back to the manager.

Reply:
Hmm, I think that could work, have pushed the changes.